cartographer_ros/cartographer_ros/offline_node_main.cc

```diff
@@ -137,6 +152,8 @@ void Run(const std::vector<string>& bag_filenames) {
     rosbag::View view(bag);
     const ::ros::Time begin_time = view.getBeginTime();
     const double duration_in_seconds = (view.getEndTime() - begin_time).toSec();
+    // The message processing loop publishes the clock.
```

Comment:
I think it is more readable if you move this down after `FinishTrajectory()` and restart before waiting for shutdown. Then you can also change the comment, i.e. say that we publish while waiting for trajectory finalization.